### ISSUE TITLE
ERROR("Non-constant 'constant_value' input to Pad would result in dynamic memory allocation")

### ISSUE THREAD

**Issue by henkten** [kraiskil/onnx2c#32]
Hello, first of all thank you for this great repository.

I have the following error when I try to export my onnx model to c using onnx2c.

[onnx2c/src/nodes/pad.cc, line 45](https://github.com/kraiskil/onnx2c/blob/master/src/nodes/pad.cc#:~:text=ERROR(%22Non%2Dconstant%20%27constant_value%27%20input%20to%20Pad%20would%20result%20in%20dynamic%20memory%20allocation%22))

Unfortunately I am a beginner in C and have no idea how to fix this error.

Here is an overview of the onnx model (exported from pytorch).

![TCN_Model-sim onnx](https://user-images.githubusercontent.com/48253223/231475689-8ae5961c-c476-487c-bede-25aab6377acf.png)

Does anyone know a way to fix this error or can help me.
I will be happy to answer any questions about the model or my setup.

Thank you very much in advance!
With kind regards
Henk

**Comment by henkten** [comment]
Hi,
thank you very much for your quick reply.

I have checked the `constant_value = inputs[2]` for my pad nodes. The image shows that the constant_value is an empty string. I suspect that is the error. 
![image](https://user-images.githubusercontent.com/48253223/231891843-4f82293b-e1ec-45c2-9e3d-54075b2cb05b.png)

I deleted the constant _value from all pad nodes and tested the model a second time using onnx2c. 
Now the code runs through this section without error.

Unfortunately, I have a new [error](https://github.com/kraiskil/onnx2c/blob/master/src/nodes/cast.cc#:~:text=ERROR(%22Unimplemented%20casting%20to%20requested%20type%22)%3B) for the cast nodes. For this error, I have no idea what to do. 
Here is a picture of the cast node. Maybe you have an idea how to fix this problem?
![image](https://user-images.githubusercontent.com/48253223/231892130-0c7f4edb-2f5d-4f2e-8950-fabbbb2c9ccd.png)

Thank you very much in advance!
With kind regards
Henk

**Comment by henkten** [comment]
I unfortunately do not understand from the issue #26  what I can change, do you have a suggestion how I can work around the error?

Sorry, i split the diagram of the simplified net above. Here is the diagram of the original mesh.
![Darts_TCN onnx](https://user-images.githubusercontent.com/48253223/232025954-72eea15f-7b99-4d86-a6b4-9ecfc156259b.png)

**Comment by kraiskil** [comment]
The problem with onnx2c `Cast` implementation is the switch starting on line https://github.com/kraiskil/onnx2c/blob/bdfe0f76b7a5674d9ea88c02684360fd615b2d1f/src/nodes/cast.cc#L28
Casts to integers never got implemented because `Cast` really doesn't sound like a useful operand on when targeting embedded systems - such a node should be optimized away.  Maybe a tool like [onnxsim](https://github.com/daquexian/onnx-simplifier) could help fold away such nodes? Also, looking at the graph, the `Dropout` sounds like a pure training node - so one that is not implemented in onnx2c.

**Comment by henkten** [comment]
With the created model from onnxsim i got this [error](https://github.com/kraiskil/onnx2c/blob/bdfe0f76b7a5674d9ea88c02684360fd615b2d1f/src/nodes/pad.cc#L45). Here a picture from the pad node.  The model does not have cast and dropout nodes anymore.
![image](https://user-images.githubusercontent.com/48253223/232342278-419f17d2-d076-4ba3-8d34-59f4ba32b4cd.png)

**Comment by kraiskil** [comment]
This is curious. Either your model is doing something really strange with the `constant_value` input tensor, or onnx2c handles this input incorrectly.

From the image I see the `constant_value` has no name, and `mode: constant` means `constant_value` should be a scalar, so I would lean towards an unimplemented feature in onnx2c. Would it be possible to post the simplified model to debug this?

**Comment by kraiskil** [comment]
A side note, I just had a look, and ONNX is having very few (1) backend test for the Pad operand. So onnx2c being non-conformant is pretty likely.

**Comment by henkten** [comment]
Sure.
here a pic from the simplified onnx-model.
![tcn_model_mod_sim onnx (1)](https://user-images.githubusercontent.com/48253223/232728484-981c2216-8645-4d9a-bdca-815752dccbf1.png)

For a better understanding of the creation of the onnx model.
The onnx model is an export of a Temporal Convolutional Network from Pytorch. Maybe that is where the error of the missing name for `constant_value` comes from.

I can upload or send you the model if you want a better overview and understanding of the modelparameter.

**Comment by kraiskil** [comment]
Yes, please upload. Having the .onnx file allows to run onnx2c under the debugger, and this helps to understand the problem.

**Comment by henkten** [comment]
I have uploaded the original Onnx, a modified Onnx with changed input and output names, and a simplified version of the modified model.

Additionally I uploaded the script I used to modify the original model.

Here is [the link to the repo](https://github.com/henkten/pytorch2onnx2c.git).

**Comment by kraiskil** [comment]
Thanks for the files. I ran the `tcn_model_mod_sim.onnx` one to have a look at what is going on.

The problem is with the 2nd input to the `Pad` node (at least the `/res_blocks.1/Pad_1` one). This `pads` input is of the format float32[64], which is not according to ONNX specifications (https://github.com/onnx/onnx/blob/main/docs/Operators.md#Pad). `pads` are supposed to tell how much to pad with. A tensor of floats sounds more like the contents of the added cells in the output padded tensor.

> pads (non-differentiable) : tensor(int64)
    Tensor of integers indicating the number of padding elements to add or remove (if negative) at the beginning and end of each axis. For 2D input tensor, it is the number of pixels. `pads` should be a 1D tensor of shape [2 * num_axes] where `num_axes` refers to the number of elements in the `axes` input or the input rank if `axes` are not provided explicitly. `pads` format should be: [x1_begin, x2_begin, ..., x1_end, x2_end,...], where xi_begin is the number of pad values added at the beginning of axis `axes[i]` and xi_end, the number of pad values added at the end of axis `axes[i]`. 

Seeing such output sounds like a bug in onnxsim. I tried to run onnxsim on the input `tcn_model.onnx`, but that would not work, since onnxsim couldn't handle the Reshape node being of too new version (i.e. it has an 'allowzero' attribute as of ONNX-14)

Latest onnx2c version now contains at least a more explicit error message on this kind of input :)

Maybe another approach here is to look back, if padding really is necessary? In such a small network, would it be feasible to ignore the edges of the input? I would guess it would result in faster inference times.

**Comment by henkten** [comment]
Thank you for your time and investigation for my model.

This is not the original model which I use final, but a simple replica to work with it quickly and easily.
Unfortunately, padding is required because the output length must be equal to the input length. 

Did I understand it correctly then that it is not possible to manually adjust the padding operations to cancel the errors before I try to convert the model with onn2c?

**Comment by kraiskil** [comment]
> Did I understand it correctly then that it is not possible to manually adjust the padding operations to cancel the errors before I try to convert the model with onn2c?

If by this you mean manually editing the .onnx file? I don't know of such a tool, but I would imagine there would be something out there to make this possible.

The problem here is the original (pytorch?) export to .onnx is creating a long chain of operations just to end up as the `pads` input to the `Pad` operand.  (i.e. the ConstantOfShape, Concat, Reshape, Slice..., Cast). All this runtime calculation just to achieve what needs to be a compile time constant (i.e. onnx2c should recognize this `pads` tensor as a constant, otherwise the model has inference time dynamic memory allocations).

`onnxsim` should be able to do this simplification. But what I understand is the output of onnxsim in your repository is invalid, so seems you found an onnxsim bug. For some strange reason, I can't even re-run onnxsim to verify this.

It does sound strange that if you in pytorch provide a constant padding tensor to the `Pad` operand, that it would create such a long chain of operands to achieve this. Maybe there is a way of simplifying the .onnx at export time?

**Comment by henkten** [comment]
Hi, thanks for the reply, yes the onnx model is a pytorch export.

I will have a look at the export. Maybe I can find a solution for the pad node there, or a simplification of the export.
